### ISSUE TITLE
(Modules) Tracking issue for `(use) crate_name::` paths without `extern crate`

### ISSUE THREAD

**Issue by Centril** [rust-lang/rust#53128]
This is a sub-tracking issue for the RFC "Clarify and streamline paths and visibility" (rust-lang/rfcs#2126)
dealing with the deprecation of `extern crate` as well as having `use crate_name::foo::bar` and `crate_name::foo::bar` Just Work™.

- [x] Implement the RFC.
   - The main blocker is that we don't yet have the macro changes finalized to make it possible to avoid needing `#[macro_use] extern crate`.
      - See https://github.com/rust-lang/rust/pull/51496
- [ ] Adjust documentation ([see instructions on forge][doc-guide])
- [x] Stabilization PR ([see instructions on forge][stabilization-guide])

[stabilization-guide]: https://forge.rust-lang.org/stabilization-guide.html
[doc-guide]: https://forge.rust-lang.org/stabilization-guide.html#updating-documentation

**Unresolved questions:**

None

**Comment by joshtriplett** [comment]
I feel that the deprecation of `extern crate` and the automatic inclusion of external crates in the prelude has gotten enough exposure and positive feedback that it's ready for us to commit to as the path we're going down. We shouldn't stabilize it until we stabilize the macro system changes, but we can handle that via a blocking concern; in the meantime, let's confirm that we have consensus.

Note that while we're still talking about the final system for paths in `use` statements, both of the proposals currently under consideration incorporate this deprecation of `extern crate` and automatic inclusion of external crates in the prelude.

@rfcbot fcp merge

**Comment by joshtriplett** [comment]
@rfcbot concern macros

We shouldn't stabilize this until we first stabilize the macro system changes needed to deprecate `#[macro_use] extern crate` as well.

**Comment by matklad** [comment]
Not really a concern, but some tradeoffs about automatic inclusion of external crates in the prelude, mostly from [internals thread](https://internals.rust-lang.org/t/relative-paths-in-rust-2018/7883/54).

In Rust 2015, there are two ways to refer to extern crate `foo`:
* add `use foo` to the top of the file and write `foo::Bar` in the module's body (special case in the root: add `extern crate foo`).
* write `::foo::Bar` in the module body

Extern crates in prelude adds a third one:
* write `foo::Bar` in the module body

This third variant is easier to type, but it gives the least amount of information to the reader.

* With the first option, two readability properties are maintained. It is possible to use text search in a single file to see where the `foo` identifier comes from. By looking at the list of imports, it is possible to say, based on the imported crates, what the module could do, without reading its source. 
* With the second option, it is immediately clear where the `foo` comes from, especially in 2018 edition, although it's not possible to scan a list of imports to get an "overview" of the module dependencies. 
* With the the third option, neither text search, not dependencies overview work. Although the set of extern crates is usually not very big, it is project specific, so, seeing a `foo::bar` in an unfamiliar code base, it might be not trivial to understand where the `foo` comes from. 


It is also seems to be the case that adding extern crates to the prelude is not required to make other module systems changes "work". Specifically, ergonimics is not regressed compared to Rust 2015: `::foo::Bar` style paths continue to work. `use foo; foo::Bar` pairs continue to work, `extern crate foo: foo::Bar` changes to `use foo; foo::Bar`. The only exception here is `std` in the root module, whose `extern crate std` is added implicitly. This is mitigated by the fact that popular types of `std` are in prelude and that `std` does not contain top-level items, so submodules are often usually imported anyway. 

It is true that some use statements might be dropped with this feature in 2018.


The original motivation for this feature was to fix path confusion: the fact that paths in use statements and everywhere else work differently. This is only a partial fix though:

* `use foo::Bar` and `foo::Bar` in fn body will resolve to the same `Bar`, but for *different* reasons: in the first case, because the path is equivalent to `::foo::Bar`, in the second case, because `foo` is in prelude. 
* This does not solve the problem of needing `self::` in use to refer to submodules and local enums, and there's [a separate](https://internals.rust-lang.org/t/relative-paths-in-rust-2018/7883/) proposal to fix that.

TL;DR

extern crates in prelude

* seem to violate "optimize for the reader" rule,
* are not necessary to make 2018 systems coherent (and could be added in the next edition),
* don't quite solve the original problem they were a solution for.

**Comment by sanmai-NL** [comment]
@matklad: your critique ultimately hinges on the outcome of #53130, no? AFAICT, the ambiguity does not appear if we choose `anchored_use_paths` there.

**Comment by matklad** [comment]
#53130 is about paths in use statements. The addition of crates to prelude affects paths outside of use imports. I.e., even with `anchored_use_path` (which is what currently is default with edition 2018 I believe), `fn main() { regex::Regex::new(".*").unwrap().expect() }` works.

**Comment by sanmai-NL** [comment]
Ah, right. I am only commenting as production user of Rust and haven’t paid close attention to the whole discussion/RFCs so please bear with me. :slightly_smiling_face: 
Why wouldn’t an obvious solution that springs me to mind, a uniform path syntax throughout (`use` statements and elsewhere), i.e. something along the lines of `anchored_paths`, work then?

**Comment by matklad** [comment]
>a uniform path syntax throughout

The devil is in the details, and the most devilish detail here is which syntax to pick to refer to external crates. The two options are

* `foo::bar` always refers to an extern crate, unless there's a `foo` name in local scope. This is achieved via adding extern crates to prelude, and this "unless" bit is what makes me feel uneasy about this approach and the core issue behind my big comment.

* `::foo::bar` always refers to an extern crate, and you have to use that leading `::` in use statements as well (currently, it is optional in `use`, and that leads to two-flavors of paths and the need for `self::`). To me personally this looks like the obviously right solution, but turns out a lot of folks really don't like the leading `::`.

**Comment by cramertj** [comment]
@rfcbot concern no_std-std

I just opened https://github.com/rust-lang/rust/issues/53166. It's not a huge deal (it's a fairly odd corner case), but I'd like to get some amount of resolution on it before starting FCP.

**Comment by SimonSapin** [comment]
Would for example `use proc_macro::TokenStream;` "just work" without `extern crate proc_macro;`? (For sysroot crates not listed explicitly with `--extern`.)

**Comment by petrochenkov** [comment]
@SimonSapin 
This depends on the choice made with regards to https://github.com/rust-lang/rust/issues/53130.

- In the "uniform path" model `use proc_macro::TokenStream` won't work unless `proc_macro` is specifically hardcoded in the compiler, but the disambiguated form `use ::proc_macro::TokenStream` will.
- In the "anchored use path" model the starting `::` is still added implicitly, so `use proc_macro::TokenStream` will work.

**Comment by eddyb** [comment]
@SimonSapin @petrochenkov So there's one more concern, that @aturon and @joshtriplett raised when discussing #52923 (that I'm not sure is written down anywhere else), around the fact that `::test` (the crate) and `self::test` (a local module) would both work if we keep the behavior you talk about, but "uniform paths" would have to choose one of:
* detect an ambiguity for any unqualified import starting with a name for which a crate can be loaded, *even when* the crate isn't importable through an unqualified path (i.e. it's not in `extern_prelude`) - this is unacceptable given how many modules are named `test` across the ecosystem (there are more examples, like `syntax` or `log`, but those are rarer)
* special-case ambiguity detection to only fire if the crate could be imported unqualified (i.e. if it's in the `extern_prelude`), making it seem inconsistent with the simple interpretation of "only one of `::x` and `self::x` must exist"

Alternatively, we can make `::crate_name` not work for any crate not in the `extern_prelude`, and still require `extern crate` / `--extern` for those crates until a better mechanism (preferably involving `Cargo.toml`) is introduced.

**Comment by joshtriplett** [comment]
@eddyb For the record, for the moment I'd favor the solution of having the `::crate_name` syntax not work for crates not named in `Cargo.toml` / provided via `--extern`. The goal of removing `extern crate` is to avoid duplication and having to list the external crate twice, but it needs listing at least once.

**Comment by SimonSapin** [comment]
I haven’t followed the details of the different path proposals, I mostly wanted to remind that there can be more crate than Cargo dependencies and `std`. If such crates still need `extern crate` to be used, then `extern crate` still need to be taught to new users.

**Comment by aturon** [comment]
@SimonSapin to clarify, @joshtriplett is talking specifically about "sysroot" crates (like `test`) -- he mentioned that passing in via `--extern` is fine. Today these crates are available without saying anything to the compiler or Cargo, and there's an opportunity to clean up this behavior a bit.

**Comment by SimonSapin** [comment]
@aturon Are you suggesting that all sysroot crates would always be specified through `--extern`? If so it seems concerning that mentioning a `::proc_macro::Something` path anywhere in the crate is all it takes to start linking against the compiler.

**Comment by joshtriplett** [comment]
@SimonSapin 
> Are you suggesting that all sysroot crates would always be specified through --extern?

Not automatically; they'd need to be specified explicitly in order to use them, via some "allow the use of this sysroot crate" option (and equivalent in `Cargo.toml`).

> If so it seems concerning that mentioning a `::proc_macro::Something` path anywhere in the crate is all it takes to start linking against the compiler.

That's exactly the concern, and we want to avoid that.

In the short term, we'll probably just continue requiring `extern crate proc_macro;`. In the long term, we'd like to have some kind of `--extern-builtin proc_macro`, and a corresponding `Cargo.toml` dependency line like `proc_macro = "builtin"`.

**Comment by Nemo157** [comment]
I just encountered another case where you still require some explicit inclusion of a dependency with this new system: crates which have a public API of only a lang item like `panic_implementation`, e.g. [`panic-abort`](https://crates.io/crates/panic-abort). Without any reference to the dependency it is not linked into the final binary so no implementation is found.

**Comment by cramertj** [comment]
@Nemo157 yup, that's not wonderful, although I can sort of justify asking for crate linkage via `extern crate` because it feels like a special feature.

**Comment by SimonSapin** [comment]
Sounds like another case that means we can’t or shouldn’t actually deprecate `extern crate`, which seems to be the premise of this issue.

**Comment by joshtriplett** [comment]
@SimonSapin The premise of this issue is not just needing it in the common case. For the several reasons observed so far, I doubt we'll be able to fully deprecate it for a while.

**Comment by SimonSapin** [comment]
That is not what the current title and description of the issue suggest.

**Comment by alexreg** [comment]
What about just `use crate_name;` to replace the case of `extern crate crate_name;` (with no other references). If there are references to things within the crate from the dependent crate, then `use crate_name;` should be linted as superfluous. That would work for me. It offers a more consistent syntax / interface to importing at least.

**Comment by Nemo157** [comment]
One thought is whether lang items such as `panic_implementation` should be imported into binary crates. Then you could have `use panic_abort::panic_implementation;` to make it obvious why it’s being imported. I’m not sure how that would work with `std`s implementations though, could it be added to the implicit prelude in binary crates only?

**Comment by SimonSapin** [comment]
`#[global_allocator]` is stable since 1.28 and does not need to be imported that way.

**Comment by eddyb** [comment]
I'd suggest `use crate_name as _;` to replace the link-only case.
This matches `use foo::Trait as _;` to access the methods without naming the trait.

**Comment by alexreg** [comment]
@eddyb Ah, that's a slightly upgrade on my suggestion, I agree!

**Comment by solb** [comment]
Clarification question: how will this change the behavior for those of us who sometimes invoke rustc directly (or via Make) instead of using Cargo?  Currently, if I have a compiled Rust library `libfoo.rlib` or `libfoo.so` in my library search path, I can ask it to be linked in simply by typing `extern crate foo;`. I find the symmetry with the `foo.rs`/`use foo;` case beautiful, and I hope we're not sacrificing it with the crate/module changes?

**Comment by joshtriplett** [comment]
On August 16, 2018 2:45:40 PM PDT, Sol Boucher <notifications@github.com> wrote:
>Clarification question: how will this change the behavior for those of
>us who sometimes invoke rustc directly (or via Make) instead of using
>Cargo?  Currently, if I have a compiled Rust library `libfoo.rlib` or
>`libfoo.so` in my library search path, I can ask it to be linked in
>simply by typing `extern crate foo;`. I find the symmetry with the
>`foo.rs`/`use foo;` case beautiful, and I hope we're not sacrificing it
>with the crate/module changes?

If you pass a crate with `--crate` then it'll show up in the extern prelude and you can just use it.

If you don't pass `--crate`, it won't show up in the extern prelude, and this new syntax won't work for it. @aturon and I have been talking about ways to improve that case, but that isn't what this issue addresses.

**Comment by eddyb** [comment]
Note that for now, implementation-wise, paths that start with `::` will load the crate from the sysroot or a library search path, regardless of `--extern` being used or not.
But *if you can*, I'd *highly suggest* replacing library search paths with `--extern`.

**Comment by Nemo157** [comment]
> I'd suggest `use crate_name as _;` to replace the link-only case.

I like this syntax, two issues when trying to use it for `panic-abort`: 

1. It's still [an unstable feature](https://github.com/rust-lang/rust/issues/48216) (not a big problem, but if `extern crate` is deprecated and this is the official recommendation it'll need stabilising first).

2. It gives an unused import error: `` warning: unused import: `panic_abort as _` ``, that's [fixed for `extern crate panic_abort as _`](https://github.com/rust-lang/rust/pull/53479), but I'm not sure how that could work for `use`, maybe it can detect when the path refers to the root of a package and assume it's always used, while continuing to warn for internal modules?

**Comment by petrochenkov** [comment]
Right now warnings are issued for `use path as _;`/`extern crate as _;` exactly like for `use path as name;`/`extern crate as name;`.
If the import is not used, then `unused_imports` is reported.
If the extern crate is not used, then `unused_extern_crates` is reported.

If extern crate or import is used in some undetectable way (e.g. for linking only), then, IMO, it should have explicit `#[allow(unused_extern_crates)]` on it and a comment explaining why this is necessary at all.

There was some discussion about it in https://github.com/rust-lang/rust/pull/42588 and `unused_extern_crates` was made warn-by-default as a result, requiring explicit annotations for link-only extern crates.
It was reverted later due to another issue (https://github.com/rust-lang/rust/pull/44825) though.

**Comment by petrochenkov** [comment]
https://github.com/rust-lang/rust/pull/53479 still looks reasonable though, since there's no other way to use a crate gensymed with `_` except for linking.
This logic can be transferred on `use crate_name as _;`, I think. `as _` is as good explicit annotation as `#[allow(unused_extern_crates)]` (but having comment explaining why this import is necessary would still be nice).

**Comment by joshtriplett** [comment]
@eddyb I had the same thought yesterday, which is why I worked on https://github.com/rust-lang/rust/pull/53479 . I do think we should stabilize `underscore_imports` as part of Rust 2018, since it fits naturally as a complimentary part of the module system updates. We do need the remaining issue fixed, though, namely that `use foo as _;` generates an unused import warning.

**Comment by aturon** [comment]
@joshtriplett 

> We shouldn't stabilize this until we first stabilize the macro system changes needed to deprecate #[macro_use] extern crate as well.

To be clear, when you say "this" do you mean specifically the deprecation? The title and description on this tracking issue are at odds over exactly what's covered. I agree re: deprecation (at least for the macro case) but feel we need to stabilize the path part of this with Rust 2018.

**Comment by joshtriplett** [comment]
@aturon Yes, sorry. We need to avoid having deprecation warnings for `extern crate` if you still need `#[macro_use] extern crate` for macros. But I think the changes to support `use` for all macros are on track at this point.

**Comment by alexreg** [comment]
@joshtriplett But `#[macro_use] extern crate` is only needed in Rust 2015, right?

**Comment by mark-i-m** [comment]
Huh? I was under the impression that you still need macro_use in 2018 (for now).

**Comment by alexreg** [comment]
@mark-i-m No. You can employ `use` syntax to import macros from other crates, even `macro_rules!` ones. It's only if you want to use `macro_rules!` internally that you need to employ `macro_use`. But the recommended way forward is to use Macros 2.0 `macro` declarations. That's how I understood it at least... someone correct me if I'm horribly mistaken!

**Comment by mark-i-m** [comment]
Woah, you are right! That's a major ergonmics improvement I'M HOME :)

**Comment by rpjohnst** [comment]
Going back to @matklad's comment:

> The only exception here is `std` in the root module, whose `extern crate std` is added implicitly. This is mitigated by the fact that popular types of `std` are in prelude and that `std` does not contain top-level items, so submodules are often usually imported anyway.

This is actually, IIRC, if not the *main* reason then definitely the *first* reason for adding extern crates to the prelude. Path confusion is fixed in principle without that, but with `std` being as common as it is these mitigations aren't enough to fix it for beginners IMO.

So the two choices in my mind are a) remove `extern crate std` from the crate root to make it match Rust 2015, which is a pain because it increases the frequency of `use std;`, or b) add all dependencies *including* `std` to the prelude, so they all feel the way `std` does in the crate root. Leaving it as is leaves path confusion un-fixed.

> `::foo::bar` always refers to an extern crate, and you have to use that leading `::` in use statements as well (currently, it is optional in `use`, and that leads to two-flavors of paths and the need for `self::`). To me personally this looks like the obviously right solution, but turns out a lot of folks really don't like the leading `::`.

Whenever this is brought up, I can't help but think back to some of the original module system threads, where two major themes were a) `use` paths to other crates and across the module hierarchy are *super* common, so absolute paths are an important default there, and b) every attempt to add more sigils to paths failed, despite the benefits they brought (and there were a lot of attempts!).

I also think the ambiguity here is overplayed. C# and Java, both notoriously verbose, have this property and it's never been an issue there. You can generally tell from context, for two reasons: top-level crate names are distinct enough and well-known enough not to be confused with local items, and paths that use them tend to be longer than paths referring to local items.

**Comment by Centril** [comment]
(There's not a ticky box here for me, but consider it ticked anyways...)

**Comment by sanmai-NL** [comment]
@rpjohnst:
> which is a pain because it increases the frequency of `use std;`

I miss a motivation for this qualification.

**Comment by joshtriplett** [comment]
@rfcbot resolved macros

The features needed for this are on track.

**Comment by aturon** [comment]
re: the no-std concern, we discussed this in the lang team meeting today and I've posted a summary comment [here](https://github.com/rust-lang/rust/issues/53166#issuecomment-419265401).

**Comment by cramertj** [comment]
@rfcbot resolve no_std-std

I'm happy with the proposed solution of a deny-by-default warning when `#![no_std]` is used, which crate authors can `allow` based on their own optional `std` feature.

**Comment by sanmai-NL** [comment]
@Centril: in the OP you mention `macro_use`. AFAIK this has been solved, one can `use` macros and https://github.com/rust-lang/rust/issues/53128#issuecomment-419210498 seems to confirm this issue has been solved. Could you update the OP for other readers?

**Comment by spearman** [comment]
Is there an opt-out for this?

Crates are not required to have the same name as their packages, so as far as I know there's really there's no way to know what's in scope by looking at either the source *or* Cargo.toml without looking at yet another Cargo.toml for the extern crate to see what [lib] is named. That sounds like a huge headache. Correct me if I'm wrong.

At least by requiring 'extern crate' to appear before use you're required to declare the crate name *somewhere* in the source code. It seems much better than having to look outside of the source tree, much less having to go look at the Cargo.toml of a completely different project.

**Comment by mark-i-m** [comment]
I've been using this feature for a while now and haven't had any problems. My guess is that the vast majority of crates use the same name and crates.io name (at least all of the ones I use do).

On the contrary, this is one of the most noticable ergonomics improvements in rust 2018 IMHO, and that's saying a lot because there are many nice features.

**Comment by mark-i-m** [comment]
Are there any specific crates that have caused you particular headache?

**Comment by spearman** [comment]
The only one that I can think of right now is the `fmod` package, for which the library is named `rfmod` (and to confuse things more the repository is named `rust-fmod`).

In other languages it's a common requirement that imports are declared before use. Haskell and Python both have `import` statements, C and C++ have `include`s. I never have found this to be 'un-ergonomic'. In fact it's very useful to be able to look at the first 10 or so lines of a single-file `main.rs` and see exactly what is being used within it. What doesn't make sense to me with this change, and I think might be confusing for beginners, is why what seems like an expected part of the language (declaring external dependencies) is being pushed into the build system (cargo)?

**Comment by sanmai-NL** [comment]
@spearman: I see the issue, and I have suffered it too. IMO, the proper solution is for crates.io and more generally Cargo registry tooling to enforce that ‘crates’ (packages) are named the same as the root module.

**Comment by rpjohnst** [comment]
> In other languages it's a common requirement that imports are declared before use.

This isn't changing. You still have `use` or fully-qualified paths; the only thing going away is the redundant `extern crate some_name` before the more-useful `use some_name::actual_item`.

**Comment by joshtriplett** [comment]
@spearman Because the build system has to know about it, and it shouldn't appear in *two* places.

Regarding fmod, and other crates that differ from their package names, I wonder if it might be worth going through those and working with their authors to transition to having the two match, for simplicity's sake alone?

@rpjohnst :+1:

**Comment by sanmai-NL** [comment]
@joshtriplett: Yes please! There’s also the difference in acceptance of `-` and `_` in names. This causes needless friction. I do hope, once everyone is on board, the name rules are actually going to be enforced rather than remedied afterwards in the future.

**Comment by spearman** [comment]
@joshtriplett I'm arguing it should appear in both places because they do different things: extern crate in a root module declares an external dependency, and the dependency entry in the cargo manifest tells the build system where to find that dependency (whether it's on crates.io, is a git repository, or a local path, etc.)

**Comment by sanmai-NL** [comment]
The `use` statement also declares a dependency, at least is does now with the more explicit anchored paths as implemented in Rust 2018 (https://github.com/rust-lang/rust/issues/44660).

**Comment by spearman** [comment]
@sanmai-NL so does [this](https://gist.github.com/rust-play/b509d2bbbe14011d764c03dbe439936f) become [this](https://gist.github.com/rust-play/470502a5cb17030ceabcdd5f8428be22)? or [this](https://gist.github.com/rust-play/9b560636241e4ff451321514b1751165)

**Comment by sanmai-NL** [comment]
The latter. It remains unambiguous though that `rand` is an external module. You can put the use statement at top level if you think that’s clearer.

**Comment by Centril** [comment]
@rust-lang/lang From what I can tell there's nothing left to do here? Shall we close this issue?

**Comment by cramertj** [comment]
@spearman It's correct to generate a warning for that code-- the use of `rand::random()` does not require `use rand;`.

**Comment by spearman** [comment]
Previously warn(unused_extern_crates) was a way to detect unused dependencies. How can I:

1) declare dependencies in the root module
2) be warned when those dependencies are not used

As it is now, cargo builds all dependencies and does not warn if they are unused.

**Comment by cramertj** [comment]
@spearman You cannot currently do that.

**Comment by alexreg** [comment]
@mark-i-m Yeah, I'd crate one. A PR to add a new lint for this probably wouldn't be too hard.

**Comment by Centril** [comment]
I believe this is done now so I'll close this out.

**Comment by mark-i-m** [comment]
@Centril Was this issue supposed to cover getting rid of extern crate altogether? Because we have not done that yet; you still need it for core, std, etc

**Comment by cramertj** [comment]
@mark-i-m Are you referring to how std-surfaced macro intrinsics aren't yet usable by their std paths?

```rust
fn main() {
    println!("{}", std::format_args!("{}", 5));
}
```

```
error[E0433]: failed to resolve: could not find `format_args` in `std`
 --> src/main.rs:2:25
  |
2 |     println!("{}", std::format_args!("{}", 5));
  |                         ^^^^^^^^^^^ could not find `format_args` in `std`

error: aborting due to previous error
```

**Comment by mark-i-m** [comment]
@cramertj Not sure. It might be related. I was thinking more about how one must do things like `extern crate core` in no_std crates. You still have to do that for sysroot crates.

**Comment by mark-i-m** [comment]
I opened #57288.

@cramertj I don't fully understand the error you brought up. Is it the same one or different? Should I create a new tracking issue for it?

**Comment by cramertj** [comment]
@mark-i-m It is a different issue. Sure, if you would that would be great!

**Comment by mark-i-m** [comment]
Done. #57289 :)
